lib/global-styles.php

```diff
@@ -339,3 +339,82 @@ function gutenberg_experimental_global_styles_register_cpt() {
 	add_action( 'wp_enqueue_scripts', 'gutenberg_experimental_global_styles_enqueue_assets' );
 	add_action( 'admin_enqueue_scripts', 'gutenberg_experimental_global_styles_enqueue_assets_editor' );
 }
+
+/**
+ * Flatten a nested Global styles config and generates the corresponding
+ * flattened CSS variables.
+ *
+ * @param  array $config Styles configuration.
+ * @return array Flattened CSS variables declaration.
+ */
+function gutenberg_get_nested_css_variables( $config ) {
+	$result = [];
+	$token = '--';
+	foreach ( $config as $key => $value ) {
+		if ( ! is_array( $value ) ) {
+			$result[ sanitize_title_with_dashes( $key ) ] = $value;
+		} else {
+			$nested_css_variables = gutenberg_get_nested_css_variables( $value );
+			foreach ( $nested_css_variables as $subkey => $subvalue ) {
+				$result[ sanitize_title_with_dashes( $key ) . $token . $subkey ] = $subvalue;
+			}
+		}
+	}
+
+	return $result;
+};
+
+/**
+ * Flatten a nested Global styles config and generates the corresponding
+ * flattened CSS variables.
+ *
+ * @param  array $styles Styles configuration.
+ * @return array Flattened CSS variables declaration.
+ */
+function gutenberg_get_css_variables( $styles = [] ) {
+	$prefix = '--wp';
+	$token = '--';
+
+	$result = [];
+	foreach ( gutenberg_get_nested_css_variables( $styles ) as $key => $value ) {
+		$result[] = $prefix . $token . $key . ':' . $value;
+	}
+	return $result;
+}
+
+/**
+ * Renders the CSS variables for blocks
+ *
+ * @param string $block_content Output of the current block.
+ * @param array $block Block Object.
+ * @return string New block output.
+ */
+function gutenberg_apply_style_attribute( $block_content, $block ) {
+	if ( isset( $block['attrs'] ) && isset( $block['attrs']['style'] ) ) {
```

Comment:
Ideally this doesn't rely on the attributes presence but on the "support flag". We need server-side awareness of support flags. cc @gziolo did you work on that?

Reply:
Not yet, the plan was to come up with something better than `supports` :) 

Anyway, `supports` works on the server and can be exposed when registered there to the client. So we should be good in that context. You just need to say we want it and we can implement everything. @aduth has some custom code that shims `classname` or `align` flags on the server. We can add the same logic for everything else.

Let me know whether we should update https://github.com/WordPress/gutenberg/blob/master/docs/rfc/block-registration.md#backward-compatibility and the part that states that `supports` is client-side only and exists only for backward compatibility.

Reply:
Yes, I'm thinking more and more that supports is actually a good thing if done right. Don't know what you'll think.

Reply:
It's probably fine to keep it as is but make it more configurable. We still need to use data that can be serialized but we shouldn't limit ourselves to `boolean`.